src/docs/arch-prefetch.dox

```diff
@@ -0,0 +1,106 @@
+onto the queue. Given a \c ref, it will queue the next set of pages if they are leaf
+pages, not already in the cache, and not already on the pre-fetch queue. To prevent
+overloading the pre-fetch queue, there is a maximum number of candidate pages allowed
+in the queue (\c WT_MAX_PREFETCH_QUEUE), and a limit on the maximum number of pages
+we can add to the queue during a pre-fetch check (\c WT_PREFETCH_QUEUE_PER_TRIGGER).
+Pages will not be added to the queue if either of these conditions are met.
+
+@subsection prefetch_algorithm_process_pages Pre-fetching Content Into the Cache
+Pre-fetch worker threads are responsible for the work of reading page contents
+into the cache. The \c __wt_prefetch_thread_run function is the entry point for a
+pre-fetch thread, and threads will continue to run while the pre-fetch queue is not
+empty. When there are no more pre-fetch entries to process, pre-fetch worker threads
+wait on a condition variable that signals when an entry has been pushed onto the
+pre-fetch queue. Pre-fetch worker threads invoke the \c __wt_page_in function which
+reads the page from disk and builds an in-memory representation.
+
+@section prefetch_api Pre-fetch API
+@subsection prefetch_api_connection Connection-Level Configuration
+\c wiredtiger_open("prefetch=(available=true,default=false)")
+
+The \c available setting determines whether the pre-fetching functionality can be
+enabled for the given database, and starts up the pool of pre-fetching threads if
+it is turned on. The \c default setting determines whether pre-fetching is turned
+on or off for all sessions belonging to the connection.
+
+@subsection prefetch_api_session Session-Level Configuration
+\c WT_CONNECTION.open_session("prefetch=(enabled=true)")
+
+The session-level configuration can be used to override the connection-level
+configuration for specific sessions if desired.
+
+@section prefetch_other Interaction With Other Components
+
+@subsection prefetch_other_cache_busyness Considering current cache busyness
+Cache busyness increases in proportion to the eviction workload and the rate of pages
+being read into the cache. It is important to consider how pre-fetch impacts cache
+utilization as it may leave the cache in a state that increases operational latency.
+Pre-fetch stops adding page references to the queue as soon as it detects that we are
+close to hitting the eviction clean trigger. It will also skip processing page
+references that are already on the queue until space in the cache clears up.
+
+@subsection prefetch_other_refs Ensuring references on the pre-fetch queue remain valid
+Page references on the pre-fetch queue may become invalid if other operations
+that modify page references are happening in parallel. Eviction is not allowed to
+operate on any page references that are on the pre-fetch queue, because it may
+split pages and free the underlying ref/s. Since eviction has no way to coordinate
```

Comment:
Eviction operates on pages residing in memory. Prefetch wants to load pages on disk to memory.
I wonder how an overlap might occur for Eviction and Prefetch to overlap. Or is it just a precaution?

Note: I don't know it this should necessarily be a part of **this** doc, I'm just curious.

Reply:
It's not just a precaution, and yes your other comment is correct.